Drop jetifier

malinajirka opened the pull request:
Merge Instructions:
1. ~Review and approve (don't merge) - https://github.com/wordpress-mobile/gutenberg-mobile/pull/4361~
2. ~Review this PR~
3. ~Merge Aztec https://github.com/wordpress-mobile/AztecEditor-Android/pull/946/files~
4. ~Merge Stories https://github.com/Automattic/stories-android/pull/717~
5. ~Merge GB PR - https://github.com/wordpress-mobile/gutenberg-mobile/pull/4361~
6. ~Merge GB PR which updates Aztec Version - it doesn't exist yet~
7. ~Update GB hash/version in this repo~
8. ~Review and Merge upgrade of test dependencies - https://github.com/wordpress-mobile/WordPress-Android/pull/16170 (technically not a blocker)~
9. ~Merge the branch in .mobile-secrets into trunk~
10. ~Update hash and branch in .configure~ (9f811ea608958fb7acbe781672b4c4c91318c5d4)
11. ~Remove "Not ready for merge" label~
12. ~Merge this PR~
13. Optional: Ping the team that they can change `android.enableJetifier=true` to `android.enableJetifier=false` in their gradle.properties or run `./gradlew configureApply`.

This PR migrates all libraries to versions which are migrated to AndroidX so we can disable jetifier on our builds. Jetifier was always intended to be used only temporarily and is planned to be removed from future versions of AGP.


List of changes
- Bumps Stetho version to 1.6.0 - the only change is that the version is migrated to AndroidX - https://github.com/facebook/stetho/releases
- Bumps ExoPlayer from 2.9.3 to 2.12.3 -> this was the most challenging task as ExoPlayer is used in both MediaPreivewFragment and Gutenberg-Mobile => we had to bump the version in both Gutenberg and MediaPreviewFragment. (This work isn't done yet, the GB part is still WIP)
- Updates WordPress utils version to a version which is migrated to AndroidX -> I bumped the version in all the modules to make the version consistent
- Adds our S3 repo to Analytics and Networking modules so the new version of WordPress utils is reachable
- Disables jetifier - updated gradle.properties and .configure
- "Manually" (using cmd tool) Jetified the Tenor library since it hasn't been updated since 2018 and I don't think it'll get updated any time soon
- Updated Aztec version with version that uses the updated version of WordPress utils

To test:
Make sure to disable Jetifier in your local gradle.properties  => Smoke test the app - anything you can think of.


1. Open Media Library
2. Open a detail of a video
3. Tap on the preview
4. Make sure the preview is displayed correctly and you can play the video
-------------------
1. Create a new post
2. Add image block
3. Select "Choose from Tenor"
4. Type "Cat"
5. Add a gif
---------------
1. Create a new post 
2. Add Cover block
3. Add Video
4. Choose a video file
5. Make sure the video preview works


## Regression Notes
1. Potential unintended areas of impact
- Video previews (media library)
- Video blocks in Gutenberg 
- Adding gifs from Tenor

2. What I did to test those areas of impact (or what existing automated tests I relied on)
The above "to test" section

3. What automated tests I added (or what prevented me from doing so)

None - this is a technical improvement which doesn't change the UX

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

malinajirka committed:
Bump Stetho to AndroidX version

malinajirka committed:
Bump exoplayer to version which supports AndroidX

malinajirka committed:
Exclude support library from exoplayer

malinajirka committed:
Make utils versions consistent across modules

malinajirka committed:
Add a8c-libs s3 repo to Analytics and Networking modules

malinajirka committed:
Disable jetifier in gradle.properties

malinajirka committed:
Add jetified aar of tenor library

malinajirka committed:
Update Event Bus in utils library to 3.3.1

malinajirka committed:
Update fluxc to event bus 3.3.1

malinajirka committed:
Update utils in all modules

malinajirka committed:
Bump aztec version

malinajirka committed:
Update utils so pinned version doesn't win over hash

Sometime when we specify version using a commithash eg.
`trunk-1ed207c03d2242b6fc3d74f9e388e9163cbc82a6` a pinned version wins and eg. version `1.22` is used instead of the
newer `trunk-1ed207c03d2242b6fc3d74f9e388e9163cbc82a6`.

malinajirka committed:
Update .configure

malinajirka committed:
Update event bus to v3.3.1

malinajirka commented on WordPress/build.gradle:
```diff
@@ -426,9 +426,12 @@ dependencies {
-    implementation "$gradle.ext.wputilsBinaryPath:$wordPressUtilsVersion"
-    testImplementation "$gradle.ext.wputilsBinaryPath:$wordPressUtilsVersion"
-    debugImplementation "$gradle.ext.wputilsBinaryPath:$wordPressUtilsVersion"
+    implementation ("$gradle.ext.wputilsBinaryPath") {
```
All three were identical, so I used only `implementation` - or am I missing something and we need all three?

I added the strict mode to prevent gradle from wrongly considering a pinned version (eg. 1.22) as more recent than a commit-hash version (eg. trunk-1ed207c03d2242b6fc3d74f9e388e9163cbc82a6). Alternative solution is to publish a new version of Utils.

oguzkocer replied:
Thanks for updating this. It should have been updated while doing the composite build but at the time I was also worried about missing something 😅

malinajirka commented on WordPress/build.gradle:
```diff
@@ -450,7 +453,7 @@ dependencies {
     }
 
-    implementation 'com.github.Tenor-Inc:tenor-android-core:0.5.1'
+    implementation (name:'tenor-android-core-jetified', ext:'aar') // Jetified Tenor Gif library
```
Tenor hasn't been updated since 2018 - it still uses the support library. [I had to jetify the AAR locally](https://developer.android.com/studio/command-line/jetifier) and include it in the repo.

malinajirka commented on WordPress/build.gradle:
```diff
@@ -461,8 +464,8 @@ dependencies {
     // Debug
-    debugImplementation 'com.facebook.stetho:stetho:1.5.1'
-    debugImplementation 'com.facebook.stetho:stetho-okhttp3:1.5.1'
+    debugImplementation 'com.facebook.stetho:stetho:1.6.0'
```
The only change in the new version is that it uses AndroidX - https://github.com/facebook/stetho/releases

malinajirka commented on WordPress/src/main/java/org/wordpress/android/ui/media/MediaPreviewFragment.java:
```diff
@@ -294,7 +293,7 @@ public void onLoadFailed(@Nullable Exception e, @Nullable Object model) {
 
     private void initializePlayer() {
-        mPlayer = ExoPlayerFactory.newSimpleInstance(requireContext());
+        mPlayer = (new SimpleExoPlayer.Builder(requireContext())).build();
```
This was the only breaking change in the new version of ExoPlayer.

malinajirka commented on build.gradle:
```diff
@@ -2,7 +2,7 @@ buildscript {
     ext.coroutinesVersion = '1.5.2'
     ext.navComponentVersion = '2.3.5'
-    ext.wordPressUtilsVersion = '2.2.0'
+    ext.wordPressUtilsVersion = 'trunk-1ed207c03d2242b6fc3d74f9e388e9163cbc82a6'
```
Updated utils version which is using AndroidX.

malinajirka committed:
Update CI

malinajirka commented on libs/analytics/WordPressAnalytics/build.gradle:
```diff
@@ -1,6 +1,12 @@
 apply plugin: 'com.android.library'
 
 repositories {
+    maven {
```
Adds S3 repository as that's where we store Utils artifacts now.

malinajirka commented on libs/editor/WordPressEditor/build.gradle:
```diff
@@ -1,5 +1,5 @@
 buildscript {
-    ext.aztecVersion = 'v1.5.1'
+    ext.aztecVersion = '946-74d200b819d457d42d405a8858665bd7d471acc9'
```
Updates aztec to a version which is using the new version of utils which is not dependant on the support library.

malinajirka commented on libs/editor/WordPressEditor/build.gradle:
```diff
@@ -69,14 +69,21 @@ dependencies {
     api ("$gradle.ext.aztecAndroidWordPressCommentsPath:$aztecVersion")
 
+    implementation ("$gradle.ext.wputilsBinaryPath") {
+        version {
```
Same as a few comments above: Uses `strictly` to workaround gradle wrong conflict resolution for pinned vs commithash versions.

malinajirka committed:
Merge branch 'develop' into drop-jetifier-2

# Conflicts:
#	build.gradle

malinajirka committed:
Update FluxC

malinajirka committed:
Update stories version

malinajirka committed:
Remove unintentional commit of byebyejetifier

malinajirka committed:
Update gutenberg version

malinajirka committed:
Merge branch 'develop' into drop-jetifier-2

# Conflicts:
#	build.gradle
#	libs/editor/WordPressEditor/build.gradle

malinajirka committed:
Update Aztec so pinned version doesn't win over commit hash version

Sometime when we specify version using a commithash eg.
`develop-1ed207c03d2242b6fc3d74f9e388e9163cbc82a6` a pinned version wins
 and eg. version `1.22` is used instead of the
newer `develop-1ed207c03d2242b6fc3d74f9e388e9163cbc82a6`.

oguzkocer reviewed (commented):
@malinajirka I started looking into this, even [reviewed and merged stories PR](https://github.com/Automattic/stories-android/pull/717) but I see that there are a few other merge instructions to go through before this PR is ready for review/testing. My suggestion would be to convert this to a draft and only review/test it when all the dependencies are updated. Although the dependency changes are minor, there are quite a few of them and testing early might be a bit risky. I'd be happy to review the PR once the dependency PRs are reviewed/merged.

As for the usage of `strictly`, it's not pretty, but I think it's warranted here. Something we are considering soon-ish is to enable [`failOnVersionConflict` for Gradle's resolution strategy](https://docs.gradle.org/current/dsl/org.gradle.api.artifacts.ResolutionStrategy.html#org.gradle.api.artifacts.ResolutionStrategy:failOnVersionConflict()), basically making all version resolutions strict. At that time we should be able to clean up the individual requirements like this. One possible "improvement" is to add the strict resolution rule for utils & aztec at the root level, but until the dependency resolution is centralized it's questionable how much of an improvement that is.

At a glance, the rest of the changes looks good to me, but I'd like to go through them in a bit more detail when the PR is ready.

malinajirka commented:
Thanks for looking into this @oguzkocer!

>  I started looking into this, even reviewed and merged stories PR but I see that there are a few other merge instructions to go through before this PR is ready for review/testing...

I believe this PR is already ready for review/testing. It is using all the final dependencies - just not the final hashes. AFAIK it's a regular process that we test against open-PRs and merge all of the PRs in a cascade when all of them are approved. Otherwise, if we merge the PRs in the sub-repos and this PR gets blocked we might end up in an untested state. Wdyt?

If you were referring to step `Merge GB PR which updates Aztec Version - it doesn't exist yet` I believe that step is technically optional - WordPress Android will use the final Aztec version either case, so you'd be testing it with the correct version. I created that step just to keep the versions consistent across repos.

I'll try, with the help of gutenberg folks, to approve the gutenberg PRs so it's as close to the final version as possible. Marking it as a draft as you suggested until they are approved.

malinajirka commented:
Update on this PR: I've talked to @hypest and we discussed that the timing of this PR isn't ideal. We are in the middle of migrating react to a newer version and a lot of the folks are AFK, so merging the PRs in gutenberg repo is a bit more tricky. This PR is not urgent so we decided to put it on hold until the react native bump is tested and merged (sometime in January/February).

ParaskP7 reviewed (commented):
👋 @malinajirka !

First of all, so many thanks for pulling all this together, it mustn't have been easy, not at all! ❤️ 🙇 🌟 

> Not Ready for Review: We decided to put this PR on hold until gutenberg gets updated to a newer version of react native.

I skimmed this PR and all the comments you and Oguz exchanged just to get an idea, but as per the description you added and after you having talked to Stefanos, I'll wait until `Gutenberg` gets update to a newer version of `React Native` before doing a full review of this PR.

PS: Please do re-request a review when I am needed. 🙏

oguzkocer commented:
> I believe this PR is already ready for review/testing. It is using all the final dependencies - just not the final hashes. AFAIK it's a regular process that we test against open-PRs and merge all of the PRs in a cascade when all of them are approved. Otherwise, if we merge the PRs in the sub-repos and this PR gets blocked we might end up in an untested state. Wdyt?

@malinajirka How would we end up in an untested state?

> If you were referring to step Merge GB PR which updates Aztec Version - it doesn't exist yet I believe that step is technically optional

Yes and no. This was a major factor, but I wanted dependencies to be reviewed/tested first as any issues in those would propagate to this PR and would require a second set of testing. It's also not marked as optional, so I interpreted as a requirement, more on that below.


> If you were referring to step Merge GB PR which updates Aztec Version - it doesn't exist yet I believe that step is technically optional - WordPress Android will use the final Aztec version either case, so you'd be testing it with the correct version. I created that step just to keep the versions consistent across repos.

gutenberg-mobile is the main consumer of the Aztec library, so in my opinion it's more important that gutenberg-mobile is updated. This trinity between Aztec, gutenberg-mobile and WPAndroid is not ideal. I actually have an item about this in my todo list, to look into how we are using Aztec and whether we could get away with using the transitive dependency we get from gutenberg-mobile. Either way we go, we are actually leaving either gutenberg-mobile or WPAndroid in an untested state, unless we methodologically update each one together. So, in short, I don't fully agree that it's an optional step even though as you said we'd end up using the updated aztec version once this PR is merged.

> Update on this PR: I've talked to @hypest and we discussed that the timing of this PR isn't ideal. We are in the middle of migrating react to a newer version and a lot of the folks are AFK, so merging the PRs in gutenberg repo is a bit more tricky. This PR is not urgent so we decided to put it on hold until the react native bump is tested and merged (sometime in January/February).

Sounds good. @hypest can we make sure we don't end up in a situation that this PR gets blocked forever. We had some instances like that and I'd like to help avoid them if I can. If it turns out the react-native version update is not happening as fast as you expect it to, can we reconsider going forward with these changes?

malinajirka commented:
> How would we end up in an untested state?

I was thinking about the following scenario:
1. We merge [the PR in gutenberg-mobile repo](https://github.com/wordpress-mobile/gutenberg-mobile/pull/4361)
2. This current PR doesn't get merged => gutenberg-mobile develop contains changes which are not in WPAndroid develop yet
3. Someone creates another change in gutenberg-mobile repo, merges it, and updates WPAndroid => the changes from step `1` are in WPAndroid even though this current PR is not reviewed/tested/merged yet

Ideally, they'd notice in step 3 that the number of changes coming from gutenberg mobile to WPAndroid is bigger than what they created, but I'd be worried they might miss it. Even if they wouldn't miss it, it'd be quite tricky to merge their change into WPAndroid without including change from step 1. 

> would require a second set of testing.

I agree that any changes there would require second set of testing, but merging the PRs in the other repos before this PR is reviewed feels even more risky as per the above scenario☝️ .

>  This trinity between Aztec, gutenberg-mobile and WPAndroid is not ideal. 

I agree, but I feel like that any dependency that is used in multiple modules have the same issue - eg. utils. Ideally we wouldn't be introducing any breaking changes in the libraries - including aztec. If we need to introduce a breaking change, it should be obvious from the version so whoever bumps the version knows that they might be introducing runtime errors and that they need to update even the dependencies.

>  Either way we go, we are actually leaving either gutenberg-mobile or WPAndroid in an untested state, unless we methodologically update each one together. 

I might be missing something, but if we update Aztec in WPAndroid -> the new version of Aztec is used even in gutenberg editor in WPAndroid -> so if you test gutenberg editor in WPAndroid you are testing it with the new Aztec version, so I wouldn't call it untested. I agree that you might be introducing a runtime crash if Aztec contains a breaking change, but it can be caught during testing. The only benefit of updating Aztec version in gutenberg mobile is that the compiler catches breaking changes => that's why I called it "optional" but to be on the safe side I included it in the merge instructions.

Anyway, I think we both understand each other. It's just a matter of a preference. I personally prefer for these kind of tricky/bigger efforts to start testing ASAP even with the risk that a change in other repos will require retesting this PR.

oguzkocer commented:
@malinajirka I think what we don't fully agree on is what type of change this is. If it was a feature change, then I mostly agree with everything you said. However, I'd consider this a tooling change and reviewing the PR is not just testing if the app is working, but verifying that it doesn't introduce new warnings/errors in the compilation process. With gutenberg-mobile having a different structure to every other Android library (it being react-native) I felt the correct approach was to do aztec -> gutenberg-mobile -> WPAndroid, however I am now doubting myself 😅

hypest commented:
> > Update on this PR: I've talked to @hypest and we discussed that the timing of this PR isn't ideal. We are in the middle of migrating react to a newer version and a lot of the folks are AFK, so merging the PRs in gutenberg repo is a bit more tricky. This PR is not urgent so we decided to put it on hold until the react native bump is tested and merged (sometime in January/February).
>
> Sounds good. @hypest can we make sure we don't end up in a situation that this PR gets blocked forever. We had some instances like that and I'd like to help avoid them if I can. If it turns out the react-native version update is not happening as fast as you expect it to, can we reconsider going forward with these changes?

The RN upgrade is now finished and with year-turn AFKs behind us, we can now resume working on this, no desire to block this. @mchowning , can you perhaps lend a hand on this one? Thanks!

mchowning commented:
> @mchowning , can you perhaps lend a hand on this one? 

👍  I'll take a look next week

mchowning commented:
> 6. Merge GB PR which updates Aztec Version - it doesn't exist yet

I believe we can cross this one off. There's been an AztecAndroid release ([1.5.3](https://github.com/wordpress-mobile/AztecEditor-Android/releases/tag/v1.5.3)] since [the AztecAndroid PR dropping Jetifier was merged](https://github.com/wordpress-mobile/AztecEditor-Android/pull/946), and that new release of AztecAndroid is in use in both [Gutenberg](https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-aztec/android/build.gradle#L12) and [WPAndroid](https://github.com/wordpress-mobile/WordPress-Android/blob/a2218f0ea74ff4ee996b6266f09b290c43bc6aa1/libs/editor/WordPressEditor/build.gradle#L2).

malinajirka committed:
Merge branch 'trunk' of github.com:wordpress-mobile/WordPress-Android into trunk

malinajirka committed:
Merge branch 'trunk' into drop-jetifier-2

# Conflicts:
#	build.gradle
#	libs/analytics/WordPressAnalytics/build.gradle
#	libs/editor/WordPressEditor/build.gradle
#	libs/editor/example/build.gradle
#	libs/networking/WordPressNetworking/build.gradle

malinajirka committed:
Update s3 repo with exoplayer 2.12.3

mchowning committed:
Merge remote-tracking branch 'origin/trunk' into drop-jetifier-2

mchowning committed:
Update exoplayer and gutenberg-mobile versions

gutenberg-mobile version bump is needed for the exoplayer version bump

mchowning commented:
Pulled in the latest changes from `trunk` and bumped the exoplayer version to 2.13.3 along with bumping the gutenberg-mobile version (to the latest code, which is using that exoplayer version). In my testing this version of exoplayer is working fine both within Gutenberg and in WPAndroid's media picker. 

Note that I also removed the gradle configuration that had us retrieving exoplayer from our jcenter-clone because we can retrieve 2.13.3 from google's maven repository.

I think this is (finally) good-to-go with respect to gutenberg-mobile.

ParaskP7 commented:
👋 @mchowning @oguzkocer !

As I am digging deeper and deeper into how everything is connected and starting to understand the overall picture, I have a question related to this.

In addition to the below main modules, which indeed got updated with `android.enableJetifier=false`:
- `react-native-aztec`
- `react-native-bridge`
- `react-native-editor`

Shouldn't all the individual note modules and their corresponding `android.enableJetifier` be updated as well (from `true` to `false`) to make sure that this work is fully complete?
- [react-native-linear-gradient](https://github.com/wordpress-mobile/react-native-linear-gradient/tree/wp-fork/android) (❓)
- [react-native-svg](https://github.com/wordpress-mobile/react-native-svg/tree/wp-fork/android) (❓)
- [react-native-video/exoplayer](https://github.com/wordpress-mobile/react-native-video/tree/wp-fork/android-exoplayer)
- [react-native-community/slider](https://github.com/wordpress-mobile/react-native-slider/tree/wp-fork/src/android)
- [react-native-get-random-values](https://github.com/wordpress-mobile/react-native-get-random-values/tree/wp-fork/android)
- [react-native-community/masted-view](https://github.com/wordpress-mobile/react-native-masked-view/tree/wp-fork/android)
- [react-native-gesture-handler](https://github.com/wordpress-mobile/react-native-gesture-handler/tree/wp-fork/android) (❓)
- [react-native-screens](https://github.com/wordpress-mobile/react-native-screens/tree/wp-fork/android) (❓)
- [react-native-safe-area-context](https://github.com/wordpress-mobile/react-native-safe-area-context/tree/wp-fork/android) (❓)
- [react-native-reanimated](https://github.com/wordpress-mobile/react-native-reanimated/tree/wp-fork/android) (❓)
- [react-native-prompt-android](https://github.com/wordpress-mobile/react-native-prompt-android/tree/wp-fork/android)
- [react-native-webview](https://github.com/wordpress-mobile/react-native-webview/tree/wp-fork/android)
- [react-native-clipboard/clipboard](https://github.com/wordpress-mobile/react-native-clipboard/tree/wp-fork/android)

PS: The 6 question mark (❓) you see on some of those node modules it because from the `gutenberg` repo itself, after building the node modules, not GitHub, I am seeing those to either:
- Have a `gradle.properties` file, but not a `gradle wrapper` to point to latest Gradle distribution. 2 of them:
  - `react-native-svg`
  - `react-native-safe-area-context`
- Not having any Gradle related files. 4 of them:
  - `react-native-linear-gradient`
  - `react-native-gesture-handler`
  - `react-native-screens`
  - `react-native-reanimated`

Any idea why that happens? 🤔

oguzkocer commented:
> Shouldn't all the individual note modules and their corresponding android.enableJetifier be updated as well (from true to false) to make sure that this work is fully complete?

@ParaskP7 That sounds correct - or, we can remove the property all together [since the default is `false`](https://developer.android.com/jetpack/androidx).

> PS: The 6 question mark (❓) you see on some of those node modules it because from the gutenberg repo itself, after building the node modules, not GitHub, I am seeing those to either:

I am sorry, I don't understand the question. Is this about the case for gb-mobile development builds after you run `npm install`? If so, I don't have any particular knowledge on what gets pulled by npm, but Gradle wrapper is not really necessary because we don't build those as standalone projects from within `node_modules`. They'll all be built by the Gradle wrapper in the `react-native-bridge` project.

The more important thing for these forked projects is that the artifact we are publishing (by Jitpack) doesn't have the jetifier enabled as that's what WPAndroid will end up using.

---

Hope that provides some clarity. If you can clarify the above question, I can try and further help. If there is a specific situation that needs to be looked at, if you can provide a branch, I can investigate it as well.

ParaskP7 commented:
👋 @oguzkocer !

> @ParaskP7 That sounds correct - or, we can remove the property all together [since the default is false](https://developer.android.com/jetpack/androidx).

👍 

> I am sorry, I don't understand the question. Is this about the case for gb-mobile development builds after you run npm install?

Yes, exactly.

> If so, I don't have any particular knowledge on what gets pulled by npm, but Gradle wrapper is not really necessary because we don't build those as standalone projects from within node_modules. They'll all be built by the Gradle wrapper in the react-native-bridge project.

Got it. So, in order for me to understand that a bit better, we don't need the `Gradle wrapper` because we don't build those as standalone project from within the `node_modules`, but then wouldn't that mean that we also don't need their corresponding and associated `gradle.properties` configuration file and thus it can be also safely ignored? 🤔 

Thus, and if my above assumption is correct, for the scope of this issue, the `android.enableJetifier` removal  on the forks, so it defaults to `false`, is not necessary, is it?

> The more important thing for these forked projects is that the artifact we are publishing (by Jitpack) doesn't have the jetifier enabled as that's what WPAndroid will end up using.

Ah, got it, so my above assumption is incorrect, we don't need the `Gradle wrapper` but we do need the `gradle.properties`. However, I do see some node modules not having the `gradle.properties` after running `npm install`, is that also expected?

I am so confused for seeing a node module like [react-native-screens](https://github.com/wordpress-mobile/react-native-screens/tree/wp-fork/android) within the `node_modules` of the `gutenberg` project, after running the `npm install`, to not having any of those Gradle related files we are discussing, while other node modules, like the [react-native-webview](https://github.com/wordpress-mobile/react-native-webview/tree/wp-fork/android), to be having both.

> Hope that provides some clarity. If you can clarify the above question, I can try and further help.

Yes, it provides some clarity, and thank you so much! 🙏  Apologies for me not being able to ask specific questions, or clarify those to the detail needed for you to answer at your best. I am still exploring, not seeing the whole picture, thus I am trying to learn as much as I can until it clicks within.

> If there is a specific situation that needs to be looked at, if you can provide a branch, I can investigate it as well.

For all the above, the `trunk` branch is enough for you to see what I am seeing. You can build from the latest `trunk` and we can compare our `note_modules` output, for all the 13 forks that interest us.

For example, below is my output:

![image](https://user-images.githubusercontent.com/9729923/154226047-8f51fc94-fc75-4c10-a9d2-f4b625643149.png)

What we should expect to be seeing is a triplet of the `gradle-wrapper.properties` + `gradle.properties` + `settings.gradle.kts` files, but as you can see not all the 13 forks have that triplet after running `npm install`. For example:
- See the `react-native-linear-gradient`, which only has one, the `settings.gradle.kts` file,
- While the `react-native-svg` has two, the `gradle.properties` + `settings.gradle.kts` files,
- And in comparison, the `react-native-video/android-exomplayer` has all three, the `gradle-wrapper.properties` + `gradle.properties` + `settings.gradle.kts` files.
- While `react-native-gesture-handler` or `react-native-screens` doesn't have any such files.

I wonder if you get the same output. 🤔

oguzkocer commented:
> Thus, and if my above assumption is correct, for the scope of this issue, the android.enableJetifier removal on the forks, so it defaults to false, is not necessary, is it?

You already answered this in your next statement, but just for the sake of avoiding confusion by others, I'll repeat that this is incorrect because what we most care about is how the artifact is published. During that process the `gradle.properties` in the project will be utilized and how the project is handled by `npm` is not relevant.

> However, I do see some node modules not having the gradle.properties after running npm install, is that also expected?

Hopefully my specific answers at the end of this comment will provide some clarity.

> Yes, it provides some clarity, and thank you so much! 🙏 Apologies for me not being able to ask specific questions, or clarify those to the detail needed for you to answer at your best. I am still exploring, not seeing the whole picture, thus I am trying to learn as much as I can until it clicks within.

No apologies necessary. I am just clearly stating when I don't understand something so that there isn't further confusion. It's not meant as a feedback for your communication, so if it came off that way, I am sorry.

> For all the above, the `trunk` branch is enough for you to see what I am seeing. You can build from the latest `trunk` and we can compare our note_modules output, for all the 13 forks that interest us.

I am a bit confused as to what that screenshot is 😓 When you say `trunk`, I assume you mean `trunk` for `WordPress-Android`, but if that's the case how are you getting those files to show up in Android Studio unless you are using composite build? I think I am going to need some instructions to produce the result you're looking for. (that's unless my replies below doesn't already answer your question)

---

> What we should expect to be seeing is a triplet of the gradle-wrapper.properties + gradle.properties + settings.gradle.kts files, but as you can see not all the 13 forks have that triplet after running npm install. For example:

I'll reply to this part assuming that you're referring to the `node_modules` folder.

> See the `react-native-linear-gradient`, which only has one, the `settings.gradle.kts` file,

[Because the other files are ignored through `.npmignore` within the `android` folder.](https://github.com/wordpress-mobile/react-native-linear-gradient/blob/wp-fork/android/.npmignore)

> While the `react-native-svg` has two, the `gradle.properties` + `settings.gradle.kts` files,

Similarly, [the Gradle wrapper files are ignored.](https://github.com/wordpress-mobile/react-native-svg/blob/wp-fork/android/.npmignore)

> And in comparison, the `react-native-video/android-exoplayer` has all three, the `gradle-wrapper.properties` + `gradle.properties` + `settings.gradle.kts` files.

[Because `android-exoplayer` is part of the packaged files](https://github.com/wordpress-mobile/react-native-video/blob/wp-fork/package.json#L47) and there is no `.npmignore` within [this directory](https://github.com/wordpress-mobile/react-native-video/tree/wp-fork/android-exoplayer).

> While `react-native-gesture-handler` or `react-native-screens` doesn't have any such files.

[In `react-native-gesture-handler` these files are not added to the package.](https://github.com/wordpress-mobile/react-native-gesture-handler/blob/wp-fork/package.json#L24-L28)

[In `react-native-screens` these files are not added to the package.](https://github.com/wordpress-mobile/react-native-screens/blob/wp-fork/package.json#L25-L27)

---

I am hoping that the npm related answers can provide you some clarity, but I want to re-iterate that how these projects are handled within the react ecosystem shouldn't have any impact on our work about the jetifier. Admittedly this was interesting to learn about and may have implications on our other work, so I guess it all worked out somehow 😅 

@ParaskP7 Let me know if you have any further questions!

ParaskP7 commented:
👋 @oguzkocer !

> You already answered this in your next statement, but just for the sake of avoiding confusion by others, I'll repeat that this is incorrect because what we most care about is how the artifact is published. During that process the gradle.properties in the project will be utilized and how the project is handled by npm is not relevant.

👍 

> Hopefully my specific answers at the end of this comment will provide some clarity.

💯 

> No apologies necessary. I am just clearly stating when I don't understand something so that there isn't further confusion. It's not meant as a feedback for your communication, so if it came off that way, I am sorry.

🤗 Nothing came off the wrong way, there is nothing to be sorry about, I didn't perceived any of that as feedback for my communication, it is just me letting you know that I know I am asking too many question and many of them are not too targeted or specific. 🤗 

> I am a bit confused as to what that screenshot is 😓 When you say trunk, I assume you mean trunk for WordPress-Android, but if that's the case how are you getting those files to show up in Android Studio unless you are using composite build? I think I am going to need some instructions to produce the result you're looking for. (that's unless my replies below doesn't already answer your question)

Ah, apologies for the confusing, I meant `trunk` for the `gutenberg` repo, not the `WordPress-Android`.

> I'll reply to this part assuming that you're referring to the node_modules folder.

Yes, indeed. 🙏 

> [Because the other files are ignored through .npmignore within the android folder.](https://github.com/wordpress-mobile/react-native-linear-gradient/blob/wp-fork/android/.npmignore)
> 
> Similarly, [the Gradle wrapper files are ignored.](https://github.com/wordpress-mobile/react-native-svg/blob/wp-fork/android/.npmignore)
> 
> [Because android-exoplayer is part of the packaged files](https://github.com/wordpress-mobile/react-native-video/blob/wp-fork/package.json#L47) and there is no .npmignore within [this directory](https://github.com/wordpress-mobile/react-native-video/tree/wp-fork/android-exoplayer).
> 
> [In react-native-gesture-handler these files are not added to the package.](https://github.com/wordpress-mobile/react-native-gesture-handler/blob/wp-fork/package.json#L24-L28)
> 
> [In react-native-screens these files are not added to the package.](https://github.com/wordpress-mobile/react-native-screens/blob/wp-fork/package.json#L25-L27)

Oh, this is no interesting, I had no idea, thanks for sharing all that configuration, this now makes so much sense to me! 🌟

> I am hoping that the npm related answers can provide you some clarity, ...

Your answers did add so much clarity Oguz, thank you so much! 🥇 

> ... , but I want to re-iterate that how these projects are handled within the react ecosystem shouldn't have any impact on our work about the jetifier.

👍 

> Admittedly this was interesting to learn about and may have implications on our other work, so I guess it all worked out somehow 😅

😅 👍 

> @ParaskP7 Let me know if you have any further questions!

One last question and call to actions from my side. 😊 

Since we do need to disable or remove the `android.enableJetifier` from all those 13 node module fork libraries, how do you recommend we approach that?

We can either:
1. Have someone (Cc @mchowning) to go through all of the node module fork libraries and disable or remove the flag, or
2. As part of the `Gradle & AGP` upgrade work that is currently in progress, I can myself pick this up and run each node module fork library through [bye-bye-jetifier](https://github.com/dipien/bye-bye-jetifier) to verify that we can do that, then disable or remove the flag.

Your thoughts? 🤔

oguzkocer commented:
> Ah, apologies for the confusing, I meant trunk for the gutenberg repo, not the WordPress-Android.

The default branch for gb-mobile is `develop`, so that was one part of the confusion. I never considered `gutenberg` repo, because we shouldn't work with that project in isolation and prefer gb-mobile. Anyhow, glad the confusion is cleared up.

> Since we do need to disable or remove the `android.enableJetifier` from all those 13 node module fork libraries, how do you recommend we approach that?

I think it makes sense to combine this with your Gradle upgrade PRs. Since you'll visit each repo to upgrade AGP & Gradle again, it'll be easier to get it done together. Ideally it'd be a separate PR, but in practice there is very little difference and I think it's better to be practical in this case considering how many projects there are.

ParaskP7 commented:
> The default branch for gb-mobile is develop, so that was one part of the confusion. I never considered gutenberg repo, because we shouldn't work with that project in isolation and prefer gb-mobile. Anyhow, glad the confusion is cleared up.

👍 

> I think it makes sense to combine this with your Gradle upgrade PRs. Since you'll visit each repo to upgrade AGP & Gradle again, it'll be easier to get it done together. Ideally it'd be a separate PR, but in practice there is very little difference and I think it's better to be practical in this case considering how many projects there are.

I agree on that, with the hope that me doing that wouldn't require any additional work or dependency updates on those node module forked libraries based on what [bye-bye-jetifier](https://github.com/dipien/bye-bye-jetifier) will report to me. 🤞 

@mchowning @malinajirka are you on-board with me doing that as part of the `Gradle & AGP` upgrade I am currently progressing with?

Question (❓): Does that mean that this PR will then depend on my `Gradle & AGP` upgrade to finish first, thus putting this on pause?

ParaskP7 commented:
👋 @malinajirka !

**UPDATE**: I am also seeing that [WordPress-Login-Flow-Android](https://github.com/wordpress-mobile/WordPress-Login-Flow-Android) is also having [Jetifier still enabled](https://github.com/wordpress-mobile/WordPress-Login-Flow-Android/blob/trunk/gradle.properties-example#L10), thus as part of my Gradle & AGP upgrade work I'll be disabling Jetifier on that repo as well.

Let me know if I shouldn't be doing that because I am not understand something? 🤔 

PS: I am also seeing that [WordPress-FluxC-Android](https://github.com/wordpress-mobile/WordPress-FluxC-Android) is also having [Jetifier still enabled](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/trunk/gradle.properties-example#L3), is that expected because of the fact that [WordPress-Android](https://github.com/wordpress-mobile/WordPress-Android) is not free of Jetifier just yet, while [woocommerce-android](https://github.com/woocommerce/woocommerce-android) is. Thus, both clients can enable/disable Jetifier as needed, but the default still remains `true` just because of `WPAndroid`? 🤔

malinajirka commented:
My goal was to remove jetifier from the main repos as I expected that to have the biggest impact. My understanding is that if the main repos don't require jetifier than all the subrepos don't require it either - assuming that they don't need it for tests or other dependencies that are not part of the aar. Or am I missing something?

> is that expected because of the fact that [WordPress-Android](https://github.com/wordpress-mobile/WordPress-Android) is not free of Jetifier just yet

Nope, it's because FluxC or Login was never part of the plan :). As mentioned above, I think in those repos it's technically just about changing the flag since they should be already jetifier free (leaving tests and example apps aside) - otherwise WCAndroid wouldn't work without jetifier. But yeah I agree that for the clarify and tidiness it's a good call to update them too. Thanks!

ParaskP7 commented:
👋 @malinajirka !

> My goal was to remove jetifier from the main repos as I expected that to have the biggest impact. My understanding is that if the main repos don't require jetifier than all the subrepos don't require it either - assuming that they don't need it for tests or other dependencies that are not part of the aar. Or am I missing something?

👍 Nope, this is my understanding as well. However, when adding composite builds to the equation, I am not that certain anymore... 🤔 

> Nope, it's because FluxC or Login was never part of the plan :).

👍 about `FluxC`.

What confused me about the `Login` library was the fact that I saw you did the Jetifier change on `Stories` and `Aztec` repos, but then when I moved to the `Login` repo, I didn't see the flag being flipped like I was expected, thus I started thinking.

> As mentioned above, I think in those repos it's technically just about changing the flag since they should be already jetifier free (leaving tests and example apps aside) - otherwise WCAndroid wouldn't work without jetifier. But yeah I agree that for the clarify and tidiness it's a good call to update them too. Thanks!

👍

PS: About `FluxC`, it seems that this repo will be the last one of them that needs to become Jetifier free (I did `Login` [already](https://github.com/wordpress-mobile/WordPress-Login-Flow-Android/pull/83)). I know that this was never part of the plan, but I am sure you have an idea about the work required on that repo as well. Can you let me know, in case you have a feeling already, how easy is to flip the flag on `FluxC`? Maybe I'll flip that flag on `FluxC` as part of the `Gradle & AGP` work I am doing atm, just like I did for other repos already. 🤔

malinajirka commented:
> What confused me about the Login library was the fact that I saw you did the Jetifier change on Stories and Aztec repos, but then when I moved to the Login repo, I didn't see the flag being flipped like I was expected, thus I started thinking.

I used the byebyejetifier tool in WCAndroid/WPAndroid and they complained about dependencies which need to be jetified in Stories, Aztec and Utils => I removed those dependencies in those repos and WCAndroid/WPAndroid stopped complaining. So I think they should be okay even with composite builds, but I haven't verified that. 
When it comes to Login library, WCAndroid/WPAndroid were not complaining about any dependencies coming from the login library => I didn't flip the flag there.
I still think that the effort will be close to zero, since AFAIU it's just about flipping the flag => no other changes are necessary afaik (assuming we don't use any dependencies which need to be jetified in integration tests which I believe we don't even have in Login library).

ParaskP7 commented:
> I used the byebyejetifier tool in WCAndroid/WPAndroid and they complained about dependencies which need to be jetified in Stories, Aztec and Utils => I removed those dependencies in those repos and WCAndroid/WPAndroid stopped complaining. So I think they should be okay even with composite builds, but I haven't verified that.

👍 

> When it comes to Login library, WCAndroid/WPAndroid were not complaining about any dependencies coming from the login library => I didn't flip the flag there.

Ah, now I get it! 👍 

> I still think that the effort will be close to zero, since AFAIU it's just about flipping the flag => no other changes are necessary afaik (assuming we don't use any dependencies which need to be jetified in integration tests which I believe we don't even have in Login library).

👍 🤞

ParaskP7 commented:
👋 @malinajirka !

**UPDATE**: Just an FYI that I am also seeing that [WordPress-MediaPicker-Android](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android) is also having [Jetifier still enabled](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android/blob/trunk/gradle.properties#L2). PS: This library [is being utilized as composite build](https://github.com/woocommerce/woocommerce-android/blob/trunk/local-builds.gradle-example#L19) on WCAndroid.

I tried to quickly disable Jetifier as part of my [Gradle & AGP upgrade](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android/pull/55) on this repo, but it wasn't as straightforward and thus I stepped back. The main issues with Jetifier on this repo are:
- The [WordPress-Utils-Android](https://github.com/wordpress-mobile/WordPress-Utils-Android) dependency, which is pointing to `2.0.0`, which uses a previous version of `EventBus` that still has Jetifier. Updating that to its latest `2.4.0` version should fix this issue.
- The [tenor-android-core](https://github.com/Tenor-Inc/tenor-android-core) dependency, which is pointing to `0.5.1`, which still has Jetifier. Creating and adding a `Jetified` AAR of this tenor library,  just like it is done in this eab1df06017b0e648ee026e51e06e0c9ce3939d2 commit on this PR should fix this issue.

Cc Shamelessly pinging @0nko here as well since I am seeing him being the last contributor to this repo.

malinajirka commented:
> Just an FYI that I am also seeing that [WordPress-MediaPicker-Android](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android) is also having [Jetifier still enabled](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android/blob/trunk/gradle.properties#L2). PS: This library [is being utilized as composite build](https://github.com/woocommerce/woocommerce-android/blob/trunk/local-builds.gradle-example#L19) on WCAndroid.

Nice find! It's very weird that the "byebye jetifier" tool didn't warn me about it 🤔 . Any ideas why that might be?

0nko commented:
@malinajirka @ParaskP7 I've created a [MediaPicker PR](https://github.com/wordpress-mobile/WordPress-MediaPicker-Android/pull/56) that removes the Jetifier as per your suggestions.

ParaskP7 commented:
👋 @malinajirka !

> Nice find! It's very weird that the "byebye jetifier" tool didn't warn me about it 🤔 . Any ideas why that might be?

😅 I actually used `bye bye jetifier` and it was the one warning me about it... 😅 

I am not sure how you applies the plugin, but I applied it on all modules and gradle configuration files, maybe you just applied it on one module or the top module, not sure how you configured and tested it. 🤔

ParaskP7 commented:
👋 @0nko !

> @malinajirka @ParaskP7 I've created a https://github.com/wordpress-mobile/WordPress-MediaPicker-Android/pull/56 that removes the Jetifier as per your suggestions.

This is awesome, thank you! 🙇

malinajirka committed:
Merge branch 'trunk' into drop-jetifier-2

# Conflicts:
#	build.gradle

malinajirka committed:
Remove enableJetifier flag from example gradle props

malinajirka committed:
Exclude support-annotations from fluxc/wellsql

malinajirka commented:
> org.wordpress:fluxc:trunk-7035f6772f4e0cb4980da0556a5762ea7904eb24 -> org.wordpress:wellsql:1.7.0 -> com.android.support:support-annotations:28.0.0

Fixed by excluding support-annotations from fluxc -> 09ff7009ec6d559590ab02816f12c907591429ca. AFAIU it's used only as a hint for static analysis tools -> shouldn't have effect at runtime.

> +---org.mockito:mockito-android:3.3.3
 Issues found: * org/mockito/android/internal/creation/AndroidTempFileLocator.class -> android.support.test.InstrumentationRegistry

IMO a false positive -> looking at the code of `AndroidTempFileLocator` it contains a reference to the support library only in a string -> it looks at `android.support.test.` and if doesn't find it it looks at `androidx.test.`.

>  +---androidx.test.espresso:espresso-contrib:3.1.0
     +---com.google.android.apps.common.testing.accessibility.framework:accessibility-test-framework:2.0
 +---androidx.test.espresso:espresso-accessibility:3.3.0-alpha05
     +---com.google.android.apps.common.testing.accessibility.framework:accessibility-test-framework:2.0

Fixed by upgrading `androidxTestVersion` and `espressoVersion` -  in a separate PR so we can easily check that it doesn't have any unintended effects.
https://github.com/wordpress-mobile/WordPress-Android/pull/16170

> Graphs to this dependency:
> +---org.robolectric:shadows-multidex:4.4
> Issues found: 
> * org/robolectric/shadows/multidex/Shadows.class -> android/support/multidex/MultiDex
> * org/robolectric/shadows/multidex/Shadows.class -> android.support.multidex.MultiDex
> * org/robolectric/shadows/multidex/ShadowMultiDex.class -> android/support/multidex/MultiDex

Fixed by upgrading `robolectric` from 4.4 -> 4.7.3 and removing unused `shadows-multidex` - in a separate PR so we can easily check it doesn't have any unintended sideffects.
https://github.com/wordpress-mobile/WordPress-Android/pull/16170


I've updated the checklist in the description.

ParaskP7 commented:
👋 @malinajirka !

Thank you for your continuous efforts on this PR and all these improvements, we are so close into merging it! 🙇 

> Fixed by excluding support-annotations from fluxc -> https://github.com/wordpress-mobile/WordPress-Android/commit/09ff7009ec6d559590ab02816f12c907591429ca. AFAIU it's used only as a hint for static analysis tools -> shouldn't have effect at runtime.

💯 

> IMO a false positive -> looking at the code of AndroidTempFileLocator it contains a reference to the support library only in a string -> it looks at android.support.test. and if doesn't find it it looks at androidx.test..

👍 Also, since this is a test related artifact, I am much more relaxed... 😅 

> Fixed by upgrading androidxTestVersion and espressoVersion - in a separate PR so we can easily check that it doesn't have any unintended effects.
https://github.com/wordpress-mobile/WordPress-Android/pull/16170

🥇 Let's work together in merging this dependant PR! 🚀 

> Fixed by upgrading robolectric from 4.4 -> 4.7.3 and removing unused shadows-multidex - in a separate PR so we can easily check it doesn't have any unintended sideffects.
https://github.com/wordpress-mobile/WordPress-Android/pull/16170

🙏 

> I've updated the checklist in the description.

🙇

AliSoftware commented:
FYI: I'm doing the `19.6` code freeze today, so I'm moving the milestone of this PR to `19.7`

malinajirka committed:
Merge branch 'trunk' into drop-jetifier-2

malinajirka committed:
Merge branch 'trunk' into drop-jetifier-2

ParaskP7 reviewed (approved):
👋 @malinajirka !

Once more, thank you so much for creating this PR, working on it and progressing it till the very end. I reviewed it once more and tested it as per the description. I also applied the [bye-bye-jetifier](https://github.com/dipien/bye-bye-jetifier) plugin one more, with its new version, the [1.2.2](https://github.com/dipien/bye-bye-jetifier/releases/tag/v1.2.2) and verified that we are now jetifier free (except of course `org.mockito:mockito-android:3.3.3` which you anyway explained why [here](https://github.com/wordpress-mobile/WordPress-Android/pull/15675#issuecomment-1078753205)).

As such, it is a 🟢 light from my side and we can progress with the remaining merge instructions (step 9 and rest).

PS: I also update the milestone to `19.8` as I would recommend we merge this next Tuesday (19/04/22), right after the `19.7` is cut, just so we get a full 2 weeks of engineering testing before the next beta. Wdyt?

malinajirka commented:
Awesome, thank you so much for all the work you've put into this! 🙇 

> I also update the milestone to 19.8 as I would recommend we merge this next Tuesday (19/04/22)

I was planning on suggesting that too 🍻!

I'll proceed with the remaining steps and merge on Tuesday/Wednesday.

malinajirka committed:
Disable jetifier on CI
